### PR TITLE
UDTs split between .h and .cpp files.

### PR DESCRIPTION
**Pull request by shplss**

### CODE REVIEW

File: Bakery.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <iostream>
+#include "LeakedObjectDetector.h"
```
Comment (matkatmusic) [inline Bakery.h:4]: Don't include what you don't use.   You're not using anything from iostream here
- [ ] revise

File: Neighbourhood.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <iostream>
+#include "LeakedObjectDetector.h"
```
Comment (matkatmusic) [inline Neighbourhood.h:4]: Don't include what you don't use.   You're not using anything from iostream here
- [ ] revise

File: RecStudio.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <iostream>
+#include "LeakedObjectDetector.h"
```
Comment (matkatmusic) [inline RecStudio.h:4]: Don't include what you don't use.   You're not using anything from iostream here
- [ ] revise

File: StudioComplex.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <iostream>
+#include "LeakedObjectDetector.h"
```
Comment (matkatmusic) [inline StudioComplex.h:4]: Don't include what you don't use.   You're not using anything from iostream here
- [ ] revise

File: Supermarket.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <iostream>
+#include <math.h>
+#include "LeakedObjectDetector.h"
```
Comment (matkatmusic) [inline Supermarket.h:5]: Don't include what you don't use.   You're not using anything from iostream or math here
- [ ] revise

File: Wrappers.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+#include "RecStudio.h"
+#include "Supermarket.h"
+#include "Bakery.h"
+#include "StudioComplex.h"
+#include "Neighbourhood.h"
+
+struct RecStudioWrapper
+{
+    RecStudioWrapper(RecStudio* ptr) : ptrToStudio(ptr) {}
+
```
Comment (matkatmusic) [inline Wrappers.h:14]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    RecStudio* ptrToStudio = nullptr;
+
+    JUCE_LEAK_DETECTOR(RecStudioWrapper)
+};
+
+struct SupermarketWrapper
+{
+    SupermarketWrapper(Supermarket* ptr) : ptrToSupermarket(ptr) {}
+
```
Comment (matkatmusic) [inline Wrappers.h:25]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    ~SupermarketWrapper();
+
+    Supermarket* ptrToSupermarket = nullptr;
+
+    JUCE_LEAK_DETECTOR(SupermarketWrapper)
+};
+
+struct ProductWrapper
+{
+    ProductWrapper(Supermarket::Product* ptr) : ptrToProduct(ptr) {}
```
Comment (matkatmusic) [inline Wrappers.h:35]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    ~ProductWrapper();
+
+    Supermarket::Product* ptrToProduct = nullptr;
+
+    JUCE_LEAK_DETECTOR(ProductWrapper)
+};
+
+struct BakeryWrapper
+{
+    BakeryWrapper(Bakery* ptr) : ptrToBakery(ptr) {}
```
Comment (matkatmusic) [inline Wrappers.h:46]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    ~BakeryWrapper();
+
+    Bakery* ptrToBakery = nullptr;
+
+    JUCE_LEAK_DETECTOR(BakeryWrapper)
+};
+
+struct BreadWrapper
+{
+    BreadWrapper(Bakery::Bread* ptr) : ptrToBread(ptr) {}
```
Comment (matkatmusic) [inline Wrappers.h:57]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    ~BreadWrapper();
+
+    Bakery::Bread* ptrToBread = nullptr;
+
+    JUCE_LEAK_DETECTOR(BreadWrapper)
+};
+
+struct StudioComplexWrapper
+{
+    StudioComplexWrapper(StudioComplex* ptr) : ptrToStudioComplex(ptr) {}
```
Comment (matkatmusic) [inline Wrappers.h:68]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    ~StudioComplexWrapper();
+
+    StudioComplex* ptrToStudioComplex = nullptr;
+
+    JUCE_LEAK_DETECTOR(StudioComplexWrapper)
+};
+
+struct NeighbourhoodWrapper
+{
+    NeighbourhoodWrapper(Neighbourhood* ptr) : ptrToNeighbourhood(ptr) {}
```
Comment (matkatmusic) [inline Wrappers.h:79]: NO IN-CLASS IMPLEMENTATION
- [ ] revise

File: Wrappers.h [hunk 13]
```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include "LeakedObjectDetector.h"
+
+#include "RecStudio.h"
+#include "Supermarket.h"
+#include "Bakery.h"
+#include "StudioComplex.h"
+#include "Neighbourhood.h"
```
Comment (matkatmusic) [inline Wrappers.h:9]: When you move your constructor implementations to the cpp file, you won't be using any of these classes in this header.
Don't include what you don't use.
- [ ] revise.

Comment (shplss) [reply Wrappers.h:9]: I did move them, but it gives me an 'unknown type' error if they are not included.

File: Wrappers.h [hunk 14]
```diff
@@ -10,7 +10,7 @@
 
 struct RecStudioWrapper
 {
-    RecStudioWrapper(RecStudio* ptr) : ptrToStudio(ptr) {}
+    RecStudioWrapper(RecStudio*);
```
Comment (matkatmusic) [inline Wrappers.h]: You aren't using any of these owned types in this header file.  They're pointers, so you can forward declare them here and include in the CPP file.  That is the entire point of this project part: learning where you need to include and where you can instead forward declare.
- [ ] revise.

File: Bakery.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    int XBread_Amount;
+    int XCake_Amount;
+    float daySales;
+    int XClient_totalProdDeliver;
+
+    Bakery() : numBagsFlour(20),
+               XBread_Amount(40),
+               XCake_Amount(5),
+               daySales(5000.0f),
+               XClient_totalProdDeliver(5) {}
```
Comment (matkatmusic) [inline Bakery.h:21]:
```
 5) NO IN-CLASS IMPLEMENTATION ALLOWED.
         the only exception is the existing Atomic.h and LeakedObjectDetector.h
```
- [ ] revise.

File: Supermarket.h [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+struct Supermarket 
+{
+    int employPerShift { 15 };
+    int numActiveChashiers { 5 };
+    float salesPerDay { 10000.0f };
+    float profitPerDay { 3000.0f };
+    int topSellerID { 0 };
+
+    Supermarket() {}
+
```
Comment (matkatmusic) [inline Supermarket.h:18]:
```
 5) NO IN-CLASS IMPLEMENTATION ALLOWED.
         the only exception is the existing Atomic.h and LeakedObjectDetector.h
```
- [ ] revise.

File: StudioComplex.h [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+ new UDT 4:
+ with 2 member functions
+ */
+
+struct StudioComplex
+{
+    RecStudio mainStudio;
+
+    StudioComplex(RecStudio rec) : mainStudio(rec) { }
+
```
Comment (matkatmusic) [inline StudioComplex.h:17]:
```
 5) NO IN-CLASS IMPLEMENTATION ALLOWED.
         the only exception is the existing Atomic.h and LeakedObjectDetector.h
```
- [ ] revise.

File: RecStudio.h [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    int sessionPerMonth;
+    float multiDayDiscount;
+    float revenewPerMonth;
+
+    RecStudio() : numMicrophones(25),
+                  numOutboardUnits(30),
+                  sessionPerMonth(10),
+                  multiDayDiscount(0.2f),
+                  revenewPerMonth(6000.0f) {}
+
```
Comment (matkatmusic) [inline RecStudio.h:22]:
```
 5) NO IN-CLASS IMPLEMENTATION ALLOWED.
         the only exception is the existing Atomic.h and LeakedObjectDetector.h
```
- [ ] revise.

File: Neighbourhood.h [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+ with 2 member functions
+ */
+
+struct Neighbourhood
+{
+    Supermarket superMarket;
+    Bakery bakery;
+
+    Neighbourhood() { }
+
```
Comment (matkatmusic) [inline Neighbourhood.h:19]:
```
 5) NO IN-CLASS IMPLEMENTATION ALLOWED.
         the only exception is the existing Atomic.h and LeakedObjectDetector.h
```
- [ ] revise.

**Pull request merged**